File: integration-test/test/PluginsSpec.js
'use strict'

const asUserWithDocumentSet = require('../support/asUserWithDocumentSet')
const asUserWithPdfDocumentSet = require('../support/asUserWithPdfDocumentSet')

describe('Plugins', function() {
  asUserWithDocumentSet('Metadata/basic.csv', function() {
    beforeEach(function() {
      this.browser.loadShortcuts('documentSet')
      this.browser.loadShortcuts('jquery')
      this.documentSet = this.browser.shortcuts.documentSet
    })

    it('should pass server, origin, apiToken and documentSetId in the plugin query string', async function() {
      const server = await this.documentSet.createViewAndServer('show-query-string', { serverUrlFromPlugin: 'https://server-url-from-plugin' })

      try {
        await this.browser.inFrame('view-app-iframe', async () => {
          // Wait for load. This plugin is loaded when the <pre> is non-empty
          await this.browser.assertExists({ xpath: '//pre[text() and string-length()>0]', wait: true })
          const text = await this.browser.getText({ css: 'pre' })
          const params = text
            .split(/[?&]/)
            .filter(s => s.length > 0)
            .reduce((h, s) => { const [ k, v ] = s.split(/=/).map(decodeURIComponent); h[k] = v; return h }, {})
          expect(params.server).to.eq('https://server-url-from-plugin')
          expect(params.origin).to.eq(process.env.OVERVIEW_URL)
          expect(params.documentSetId).to.match(/\d+/)
          expect(params.apiToken).to.match(/[a-z0-9A-Z]+/)
        })
      } finally {
        await server.close()
      }
    })

    describe('with a plugin that calls setRightPane', async function() {
      beforeEach(async function() {
        this.server = await this.documentSet.createViewAndServer('right-pane')
      })

      afterEach(async function() {
        if (this.server) await this.server.close()
      })

      it('should create a right pane', async function() {
        const b = this.browser

        await b.assertNotExists({ id: 'tree-app-vertical-split-2' }) // it's invisible

        // wait for load
        await b.inFrame('view-app-iframe', async () => {
          await b.assertExists({ css: 'body.loaded', wait: 'pageLoad' })
          await b.click({ button: 'Set Right Pane' })
        })

        await b.assertExists({ id: 'tree-app-vertical-split-2', wait: true }) // wait for animation
        await b.click({ css: '#tree-app-vertical-split-2 button' })
        await b.sleep(1000) // for animation
        await b.assertExists({ id: 'view-app-right-pane-iframe' })

        const url = await b.inFrame('view-app-right-pane-iframe', async () => {
          return await b.execute(function() { return window.location.href })
        })
        expect(url).to.contain('?placement=right-pane')
      })

      it('should delete the right pane when deleting the view', async function() {
        const b = this.browser

        // 1. create the right pane
        await b.inFrame('view-app-iframe', async () => {
          await b.assertExists({ css: 'body.loaded', wait: 'pageLoad' })
          await b.click({ button: 'Set Right Pane' })
        })

        await b.assertExists({ css: '#tree-app-vertical-split-2 button', wait: true }) // wait for animation

        // 2. delete the view
        await b.shortcuts.documentSet.destroyView('right-pane')

        // Check the pane is gone
        await b.assertNotExists({ css: '#tree-app-vertical-split-2 button', wait: true })
        await b.assertNotExists({ id: 'view-app-right-pane-iframe', wait: true })
      })
    })

    describe('with a plugin that calls setModalDialog', async function() {
      beforeEach(async function() {
        this.server = await this.documentSet.createViewAndServer('modal-dialog')
      })

      afterEach(async function() {
        if (this.server) await this.server.close()
      })

      it('should create and close a modal dialog', async function() {
        const b = this.browser

        await b.assertNotExists({ id: 'view-app-modal-dialog' })

        await b.inFrame('view-app-iframe', async () => {
          // wait for load
          await b.assertExists({ css: 'body.loaded', wait: 'pageLoad' })
          await b.click({ button: 'Set Modal Dialog' })
        })

        await b.assertExists({ id: 'view-app-modal-dialog', wait: true })

        await b.inFrame('view-app-modal-dialog-iframe', async () => {
          await b.click({ button: 'Set Modal Dialog to Null', wait: 'pageLoad' })
        })

        await b.assertNotExists({ id: 'view-app-modal-dialog' })
      })

      it('should send messages from one plugin to another', async function() {
        const b = this.browser

        await b.inFrame('view-app-iframe', async () => {
          // wait for load
          await b.assertExists({ css: 'body.loaded', wait: 'pageLoad' })
          await b.click({ button: 'Set Modal Dialog' })
        })

        await b.assertExists({ id: 'view-app-modal-dialog', wait: true })

        await b.inFrame('view-app-modal-dialog-iframe', async () => {
          await b.click({ button: 'Send Message', wait: 'pageLoad' })
        })

        await b.inFrame('view-app-iframe', async () => {
          await b.assertExists({ tag: 'pre', contains: '{"This is":"a message"}', wait: true })
        })
      })
    })

    describe('with a plugin that calls setViewFilter', async function() {
      beforeEach(async function() {
        this.server = await this.documentSet.createViewAndServer('view-filter')
      })

      afterEach(async function() {
        if (this.server) await this.server.close()
      })

      it('should allow filtering by view', async function() {
        const b = this.browser

        // Wait for iframe contents to load and call setViewFilter
        await b.click({ tag: 'a', contains: 'view-filter placeholder', wait: 'pageLoad' })
        await b.click({ tag: 'span', contains: 'VF-Foo' })
        await this.documentSet.waitUntilDocumentListLoaded()
        expect(this.server.lastRequestUrl.path).to.match(/^\/filter\/01010101\?/)
        expect(this.server.lastRequestUrl.query.apiToken || '').to.match(/[a-z0-9]+/)
        expect(this.server.lastRequestUrl.query.ids).to.eq('foo')
        expect(this.server.lastRequestUrl.query.operation).to.eq('any')
        const text = await(b.getText('#document-list ul.documents'))
        expect(text).not.to.match(/First/)
        expect(text).to.match(/Second/)
        expect(text).not.to.match(/Third/)
      })

      it('should allow setViewFilterChoices', async function() {
        const b = this.browser

        await b.inFrame('view-app-iframe', async () => {
          // Wait for iframe contents to load
          await b.click({ button: 'setViewFilterChoices', wait: 'pageLoad' })
          await b.sleep(100) // make sure postMessage() goes through. TODO notify from plugin?
        })

        await b.click({ tag: 'a', contains: 'view-filter placeholder' })
        await b.click({ tag: 'span', contains: 'VF-Foo2' }) // assert it exists, really
      })

      it('should remove the ViewFilter when deleting the plugin', async function() {
        await this.browser.shortcuts.documentSet.destroyView('view-filter')
        await this.browser.assertNotExists({ tag: 'a', contains: 'view-filter placeholder', wait: true })
      })

      it('should allow setViewFilterSelection', async function() {
        const b = this.browser

        await b.inFrame('view-app-iframe', async () => {
          // Wait for iframe contents to load
          await b.click({ button: 'setViewFilterChoices', wait: 'pageLoad' })
          await b.click({ button: 'setViewFilterSelection' })
          await b.sleep(100) // make sure postMessage() goes through. TODO notify from plugin?
        })

        await this.documentSet.waitUntilDocumentListLoaded()
        expect(this.server.lastRequestUrl.path).to.match(/^\/filter\/01010101\?/) // the document list reloaded
        await b.assertExists({ tag: 'a', contains: 'VF-Foo2' }) // the tag was selected in the ViewFilter
      })
    })

    describe('with a plugin that calls setDocumentDetailLink', function() {
      beforeEach(async function() {
        this.server = await this.documentSet.createViewAndServer('view-document-detail-links')

        await this.browser.shortcuts.documentSet.openDocumentFromList('First')

        this.clickViewButton = async function(name) {
          const b = this.browser
          await b.inFrame('view-app-iframe', async () => {
            await b.assertExists({ css: 'body.loaded', wait: 'pageLoad' })
            await b.click({ button: name })
          })
        }
      })

      afterEach(async function() {
        if (this.server) await this.server.close()
      })

      it('should add the given link', async function() {
        const b = this.browser

        await this.clickViewButton("setUrl(foo)")
        await b.assertExists({ link: 'Text foo', wait: true })
      })

      it('should show the link even after page refresh', async function() {
        const b = this.browser

        await this.clickViewButton("setUrl(foo)")
        await b.assertExists({ link: 'Text foo', wait: true }) // wait for it to appear

        await b.refresh()
        await this.documentSet.waitUntilStable()
        // again, open a document
        await this.documentSet.openDocumentFromList('First')
        await b.assertExists({ link: 'Text foo', wait: true })
      })

      it('not create a duplicate link (to the same URL)', async function() {
        const b = this.browser

        await this.clickViewButton("setUrl(foo)")
        await b.assertExists({ link: 'Text foo', wait: true }) // wait for it to appear

        // waitUntilAjaxComplete: make sure we have 0 ajax requests (for next stuff)
        await b.shortcuts.jquery.listenForAjaxComplete()
        await this.documentSet.createCustomView('another view', `http://${this.server.hostname}:3333`)
        await b.shortcuts.jquery.waitUntilAjaxComplete()

        // Add the same URL again. Wait for it to complete. (That's an ajax request;
        // aren't you glad we're sure there were zero before?)
        await b.shortcuts.jquery.listenForAjaxComplete()
        await this.clickViewButton("setUrl(foo, foo with different text)")
        await b.shortcuts.jquery.waitUntilAjaxComplete()
        // Ajax is done. That means the new link has been saved to the server
        // and rendering is definitely finished. But since the URL is the same,
        // we expect nothing else to be rendered
        await b.assertNotExists({ link: 'foo with different text' })
      })

      it('should open the popup', async function() {
        const b = this.browser

        await this.clickViewButton("setUrl(foo)")
        await b.click({ link: 'Text foo', wait: true }) // wait for it to appear

        await b.assertExists({ css: 'iframe#view-document-detail', wait: 'fast' }) // wait for iframe
        const url = await b.inFrame('view-document-detail', async () => {
          return await this.browser.execute(function() { return window.location.href })
        })
        expect(url).to.match(/\?documentId=\d+/)
        expect(url).to.match(/&foo=foo/)
      })

      it('should remove the link when deleting the View', async function() {
        const b = this.browser
        await this.clickViewButton('setUrl(foo)')
        await b.assertExists({ link: 'Text foo', wait: true })
        await b.shortcuts.documentSet.destroyView('view-document-detail-links')
        await b.assertNotExists({ link: 'Text foo', wait: 'pageLoad' })
      })
    })

    describe('with a plugin that calls setViewTitle', function() {
      beforeEach(async function() {
        this.server = await this.documentSet.createViewAndServer('set-view-title')
        this.clickViewButton = async function(name) {
          const b = this.browser
          await b.inFrame('view-app-iframe', async () => {
            await b.assertExists({ css: 'body.loaded', wait: 'pageLoad' })
            await b.click({ button: name })
          })
        }
      })

      afterEach(async function() {
        if (this.server) await this.server.close()
      })

      it('should set the view title', async function() {
        const b = this.browser

        await this.clickViewButton("Set title to new-title")
        await b.assertExists({ tag: 'li', className: 'view', contains: 'new-title', wait: 'fast' })
      })

      it('should preserve the view title across page refresh', async function() {
        const b = this.browser

        await this.clickViewButton("Set title to new-title")
        await b.assertExists({ tag: 'li', className: 'view', contains: 'new-title', wait: 'fast' })

        await b.refresh()
        await this.documentSet.waitUntilStable()
        await b.assertExists({ tag: 'li', className: 'view', contains: 'new-title', wait: 'fast' })
      })
    })
  })

  asUserWithPdfDocumentSet('PdfNotes', function() {
    beforeEach(function() {
      this.browser.loadShortcuts('documentSet')
      this.browser.loadShortcuts('jquery')
      this.browser.loadShortcuts('pdfNotes')
      this.documentSet = this.browser.shortcuts.documentSet
    })

    describe('with a plugin that interacts with PdfNotes', function() {
      beforeEach(async function() {
        this.server = await this.documentSet.createViewAndServer('pdf-notes')
        this.clickViewButton = async function(name) {
          const b = this.browser
          await b.inFrame('view-app-iframe', async () => {
            await b.assertExists({ css: 'body.loaded', wait: 'pageLoad' })
            await b.click({ button: name })
          })
        }
      })

      afterEach(async function() {
        if (this.server) await this.server.close()
      })

      it('should begin PdfNote creation', async function() {
        const b = this.browser

        await b.shortcuts.documentSet.openDocumentFromList('doc1.pdf')
        // opening document takes 1s to animate: assume that quells all races

        await this.clickViewButton('Create PDF Note')
        await b.inFrame('document-contents', async () => {
          await b.assertExists({ css: 'button.addNote.toggled', wait: 'pageLoad' })
        })
      })

      it('should send document.pdfNotes to the plugin', async function() {
        const b = this.browser

        await b.shortcuts.documentSet.openDocumentFromList('doc1.pdf')
        // opening document takes 1s to animate: assume that quells all races

        await b.shortcuts.pdfNotes.createNote('PluginsSpec note test')

        await b.inFrame('view-app-iframe', async () => {
          // Wait for notify:document message to reach iframe.
          // Iframe JS will set pre.last-message to JSON.stringify(document)
          // We'll just look for a slice of the JSON
          await b.assertExists({ tag: 'pre', contains: 'PluginsSpec note test', wait: true })
        })
      })
    })
  })
})

File: integration-test/spec/plugins_spec.rb
#!/usr/bin/env ruby

require './spec/spec_helper'

describe 'Plugins' do
  def extra_session_helpers; [ SessionHelpers::PdfNotesHelper ]; end

  before do
    @user = admin_session.create_test_user
    page.log_in_as(@user)
  end

  after do
    admin_session.destroy_test_user(@user)
  end

  def click_view_button(text)
    page.within_frame('view-app-iframe') do
      page.assert_selector('body.loaded', wait: WAIT_LOAD) # wait for iframe to load
      page.click_button(text)
    end
  end

  describe 'with a CSV' do
    before do
      page.create_document_set_from_csv('files/plugins-spec/basic.csv')
    end

    it 'should pass server, origin, apiToken and documentSetId in the plugin query string' do
      page.with_mock_plugin_and_view('show-query-string', server: 'https://server') do
        page.within_frame('view-app-iframe') do
          page.assert_selector('pre', text: /./, wait: WAIT_LOAD) # wait for <pre> to be non-empty
          text = page.find('pre').text
          params = text
            .split(/[?&]/).reject(&:empty?)                      # Array of key=value (URI-encoded)
            .map{ |s| s.split(/=/).map{ |v| CGI::unescape(v) } } # Array of [ key, value ] (unencoded)
            .to_h                                                # Hash of { "key" => "value" }
          assert_equal('https://server', params['server'])
          assert_match(/\A\d+\Z/, params['documentSetId'])
          assert_equal(OVERVIEW_URL, params['origin'])
          assert_match(/\A[a-z0-9A-Z]+\Z/, params['apiToken'])
        end
      end
    end

    describe 'with a plugin that calls setRightPane' do
      it 'should create a right pane' do
        page.with_mock_plugin_and_view('right-pane') do
          page.assert_no_selector('#tree-app-vertical-split-2') # vertical-split-2 is invisible by default
          click_view_button('Set Right Pane')

          page.assert_selector('#tree-app-vertical-split-2', wait: WAIT_FAST) # wait for divider to appear
          page.waiting_for_css_transitions_on_selector('#tree-app-right-pane') do
            page.find('#tree-app-vertical-split-2 button').click
          end
          page.assert_selector('#view-app-right-pane-iframe') # assert frame exists
          src = page.find('#view-app-right-pane-iframe')[:src]
          assert_match(/\?placement=right-pane/, src)
        end
      end

      it 'should delete the right pane when deleting the view' do
        page.with_mock_plugin_and_view('right-pane') do
          click_view_button('Set Right Pane')

          page.assert_selector('#tree-app-vertical-split-2', wait: WAIT_FAST) # wait for divider to appear
          page.delete_current_view

          page.assert_no_selector('#tree-app-vertical-split-2') # divider should be gone immediately
          page.assert_no_selector('#view-app-right-pane-iframe') # iframe should be gone immediately
        end
      end
    end

    it 'should create and close a modal dialog with setModalDialog' do
      page.with_mock_plugin_and_view('modal-dialog') do
        click_view_button('Set Modal Dialog')

        page.assert_selector('#view-app-modal-dialog', wait: WAIT_FAST) # wait for modal to appear

        page.within_frame('view-app-modal-dialog-iframe') do
          page.assert_selector('body.loaded', wait: WAIT_LOAD) # wait for page to load
          page.click_button('Set Modal Dialog to Null')
        end

        page.assert_no_selector('#view-app-modal-dialog', wait: WAIT_FAST) # wait for modal to disappear
      end
    end

    it 'should send messages from one plugin to another' do
      page.with_mock_plugin_and_view('modal-dialog') do
        click_view_button('Set Modal Dialog')

        page.assert_selector('#view-app-modal-dialog', wait: WAIT_FAST) # wait for modal to appear
        page.within_frame('view-app-modal-dialog-iframe') do
          page.assert_selector('body.loaded', wait: WAIT_LOAD) # wait for modal to load
          page.click_button('Send Message')
        end

        ## Close modal, so we can get back to the other frame
        ##(unneeded, for some reason: semitransparency, perhaps?)
        #page.find('#view-app-modal-dialog').click(x: 1, y: 1) # click outside dialog to close it
        #page.assert_no_selector('#view-app-modal-dialog')

        page.within_frame('view-app-iframe') do
          page.assert_selector('pre', text: '{"This is":"a message"}', wait: WAIT_FAST) # wait for message+JS
        end
      end
    end

    describe 'with a plugin that calls setViewFilter' do
      before do
        @last_request = nil
        @server = MockPlugin.create_server('view-filter', 3333)
        @server.config[:BindAddress] = '0.0.0.0'
        @server.mount_proc '/filter/01010101' do |req, res|
          @last_request = req
          res.status = 200
          res['Content-Type'] = 'application/octet-stream'
          res['Cache-Control'] = 'no-cache'
          res.body = [ '01010101' ].pack('B*')
        end

        @thread = MockPlugin.start_server_in_thread(@server)

        page.create_custom_view({ name: 'mock plugin', url: "http://#{Socket.gethostname}:3333" })
      end

      after do
        MockPlugin.stop_server_and_join_thread(@server, @thread)
      end

      it 'should allow filtering by view' do
        page.click_link('view-filter placeholder', wait: WAIT_LOAD) # wait for plugin to load and setViewFilter
        page.find('.document-filter-popup .name', text: 'VF-Foo').click
        page.assert_selector('#document-list-title h3', text: 'one document', wait: WAIT_LOAD) # wait for search
        page.assert_no_selector('#document-list h3', text: 'First')
        page.assert_selector('#document-list h3', text: 'Second')
        page.assert_no_selector('#document-list h3', text: 'Third')

        assert_equal('/filter/01010101', @last_request.path)
        assert_match(/\A[a-z0-9]+\Z/, @last_request.query['apiToken'])
        assert_equal('foo', @last_request.query['ids'])
        assert_equal('any', @last_request.query['operation'])
      end

      it 'should allow setViewFilterChoices' do
        click_view_button('setViewFilterChoices')
        sleep(0.1) # make sure postMessage() goes through. TODO better notification mechanism?

        page.click_link('view-filter placeholder')
        page.assert_selector('.document-filter-popup .name', text: 'VF-Foo2')
      end

      it 'should remove the ViewFilter when deleting the view' do
        page.delete_current_view
        page.assert_no_selector('a', text: 'view-filter placeholder', wait: WAIT_LOAD) # wait for view to disappear
      end

      it 'should allow setViewFilterSelection' do
        click_view_button('setViewFilterSelection')

        page.assert_selector('#document-list-title h3', text: 'one document', wait: WAIT_LOAD) # wait for search
        assert_equal('/filter/01010101', @last_request.path) # the search happened
        page.assert_selector('a', text: 'selectedOneHtml,VF-Foo') # the tag was selected
      end
    end

    describe 'with a plugin that calls setDocumentDetailLink' do
      before do
        page.open_document_in_list_with_name('First')
      end

      it 'should add a link' do
        page.with_mock_plugin_and_view('view-document-detail-links') do
          click_view_button('setUrl(foo)')
          page.assert_selector('a', text: 'Text foo', wait: WAIT_FAST) # wait for link to appear
        end
      end

      it 'should show the link even after page refresh' do
        page.with_mock_plugin_and_view('view-document-detail-links') do
          click_view_button('setUrl(foo)')

          # Reload page and navigate back to a document
          page.refresh
          page.assert_selector('#document-list-title h3', text: '3 documents') # wait for page reload
          page.open_document_in_list_with_name('First')

          page.assert_selector('a', text: 'Text foo')
        end
      end

      it 'should not create a duplicate link (to the same URL)' do
        page.with_mock_plugin_and_view('view-document-detail-links') do
          click_view_button('setUrl(foo)')

          # Create another view to the same thing, and click its button
          page.create_custom_view(name: 'another view', url: 'http://localhost:3333')
          click_view_button('setUrl(foo, foo with different text)')

          page.assert_no_selector('.transaction-queue-communicating', wait: WAIT_LOAD) # wait for AJAX

          # Since the URL is the same, we expect the second link to never appear
          page.assert_no_selector('a', text: 'foo with different text')
        end
      end

      it 'should open the popup' do
        page.with_mock_plugin_and_view('view-document-detail-links') do
          click_view_button('setUrl(foo)')
          page.click_link('Text foo', wait: WAIT_LOAD) # wait for link to appear

          page.assert_selector('iframe#view-document-detail', wait: WAIT_FAST) # wait for click to create iframe
          src = page.find('iframe#view-document-detail')[:src]
          assert_match(/\?documentId=\d+/, src)
          assert_match(/&foo=foo/, src)
        end
      end

      it 'should remove the link when deleting the view' do
        page.with_mock_plugin_and_view('view-document-detail-links') do
          click_view_button('setUrl(foo)')
          page.assert_selector('a', text: 'Text foo') # wait for link to appear, so we can check it disappears later
          page.delete_current_view
          page.assert_no_selector('a', text: 'Text foo')
        end
      end
    end

    it 'should change the View title with setViewTitle' do
      page.with_mock_plugin_and_view('set-view-title') do
        click_view_button('Set title to new-title')
        page.assert_selector('.view.active .title', text: 'new-title', wait: WAIT_LOAD) # wait for title to be saved
        page.assert_no_selector('.transaction-queue-communicating', wait: WAIT_LOAD) # wait for AJAX to finish

        # Refresh the page and verify the title is still there
        page.refresh
        page.assert_selector('.view.active .title', text: 'new-title', wait: WAIT_LOAD) # wait for page load
      end
    end
  end

  describe 'with a plugin that interacts with PdfNotes' do
    before do
      page.create_document_set_from_pdfs_in_folder('files/pdf-notes-spec')
      page.open_document_in_list_with_name('doc1.pdf')
      page.wait_for_pdf_load(/first.*PDF/)
    end

    it 'should begin PdfNote creation' do
      page.with_mock_plugin_and_view('pdf-notes') do
        click_view_button('Create PDF Note')
        page.within_frame('document-contents') do
          page.assert_selector('button.addNote.toggled', wait: WAIT_FAST) # wait for postMessage
        end
      end
    end

    it 'should send document.pdfNotes to the plugin' do
      page.with_mock_plugin_and_view('pdf-notes') do
        page.create_pdf_note('first', 'FOO')
        page.within_frame('view-app-iframe') do
          page.assert_selector('pre', text: 'FOO', wait: WAIT_FAST) # wait for postMessage
        end
      end
    end

    it 'should allow goToPdfNote' do
      page.with_mock_plugin_and_view('pdf-notes') do
        page.create_pdf_note('first', 'FOO') # will send note to plugin

        page.within_frame('view-app-iframe') do
          page.assert_selector('pre', text: /FOO/, wait: WAIT_FAST) # wait for postMessage
        end

        click_view_button('Go To Last PDF Note')

        page.within_frame('document-contents') do
          page.assert_selector('.editNoteTool', wait: WAIT_FAST) # wait for postMessage
        end
      end
    end
  end
end
